tracing: add generic infrastructure for collection of tracing data

wentongwu opened the pull request:
1.  First, this commit adds user interface in tracing_format.h which can
    trace both string format and data format packet. Second, it adds
    method both for asynchronous and synchronous way. For asynchronous
    method, tracing packet will be buffered in tracing buffer first,
    tracing thread will output the stream data with the help of tracing
    backend when tracing thread get scheduled. Third, it adds UART and
    USB tracing backend for asynchronous tracing method, and adds POSIX
    tracing backend for synchronous tracing way. Also it can receive
    command from host to dynamically enable and disable tracing to have
    host capture tracing data conveniently.
2.  Add script to capture tracing stream data with UART backend. This
    script is developed based on pyserial, so install it correctly
    before using the script.
3.  Add script for usb backend to receive tracing stream data. This
    script, trace_capture_usb.py, is based on pyusb, so install it
    correctly before using the script.
4.  Add test case for tracing generic infrastructure with the help
    of added TRACING_TEST which can be used to customize tracing
    packet format (currently supporting string format and data
    format, there are kinds of data format, so far only enable CTF
    to test data format) for testing purpose. Also we add test
    cases for currently supported backends (UART, USB and POSIX).

wentongwu commented:
@mped-oticon @fnde-ot could you please take a look this patch? Thanks a lot

mped-oticon commented on subsys/debug/tracing/tracing_backend_usb.c:
```diff
@@ -0,0 +1,178 @@
+	struct usb_ep_descriptor if0_out_ep;
+} __packed;
+
+static atomic_t transfer_state;
```
Since this is busy-polled between differing contexts, you may want `volatile`.

wentongwu replied:
Thanks

mped-oticon commented on subsys/debug/Kconfig:
```diff
@@ -43,6 +45,91 @@ config TRACING
+	  Enable asynchronous tracing. This will buffer all the tracing
+	  packets to the ring buffer first, tracing thread will try to
+	  put as much data in ring buffer as possible when tracing thread
+	  get scheduled.
```
what is the policy in case of overflow in the ring buffer? might be nice to write as part of help text.

wentongwu replied:
@mped-oticon currently for overflow case, just increase the dropped number. Basically I will add backend api(for usb backend, add one more endpoint; for uart backend I still considering it, maybe one more uart or console needed) specially for overflow message output so that it doesn't contaminate the normal stream data, I'm considering it since code developing, but not find a good way, I will enhance it after this PR merged.

mped-oticon commented on subsys/debug/Kconfig:
```diff
@@ -43,6 +45,91 @@ config TRACING
+config TRACING_SYNC
+	bool "Synchronous Tracing"
+	help
+	  Enable synchronous tracing. This requires very little latency
```
"This requires the backend to be very low-latency"

wentongwu replied:
@mped-oticon thanks

mped-oticon commented on subsys/debug/Kconfig:
```diff
@@ -3,6 +3,8 @@
 # Copyright (c) 2015 Wind River Systems, Inc.
 # SPDX-License-Identifier: Apache-2.0
 
+DT_CHOSEN_Z_CONSOLE := zephyr,console
```
Does this cause inclusion of console stuff? (I am unaware, somebody else comment). 
If so, it should be made dependent on this tracing, as I can imagine not everybody forcibly wants console or device tree.

wentongwu replied:
I just use it to get uart name which is dedicated to tracing.
 
config TRACING_BACKEND_UART_NAME
        string "Device Name of UART Device for UART backend"
        default "$(dt_chosen_label,$(DT_CHOSEN_Z_CONSOLE))" if HAS_DTS
        default "UART_0"
        depends on TRACING_BACKEND_UART
        help
          This option specifies the name of UART device to be used for
          tracing backend.

mped-oticon commented on subsys/debug/tracing/CMakeLists.txt:
```diff
@@ -1,27 +1,42 @@
+  CONFIG_TRACING
+  tracing_buffer.c
+  tracing_core.c
+  tracing_format.c
```
CONFIG_TRACING is the most generic tracing kconfig. It will be defined for Segger and CTF and this.
However do we want to force the files upon people using Segger or CTF?
Solution: Specialize this kconfig

wentongwu replied:
@mped-oticon is CONFIG_TRACING_GENERIC good for you?

mped-oticon replied:
@wentongwu yep

mped-oticon commented on subsys/debug/tracing/CMakeLists.txt:
```diff
@@ -1,27 +1,42 @@
+
+zephyr_sources_ifdef(
+  CONFIG_TRACING_BACKEND_USB
+  tracing_backend_usb.c
```
Fine, this is specialized well.

mped-oticon reviewed (commented):
Looks very good, just a few comments.

mped-oticon reviewed (approved):
LGTM

wentongwu commented:
@nashif @nordic-krch could you please take a look this patch? Thanks

nashif reviewed (commented):
> First, this PR adds user interface in tracing_format.h which can trace
both string format and data format packet. Second, it adds method both
for asynchronous and synchronous way. For asynchronous method, tracing
packet will be buffered in tracing buffer first, tracing thread will
output the stream data with the help of tracing backend when tracing
thread get scheduled. Third, it adds UART and USB tracing backend for
asynchronous tracing method. Also it can receive command from host to
dynamically enable and disable tracing to have host capture tracing
data conveniently.

This is the first commit, but it is talking about the complete PR, please keep the commit messages to what the commit is doing and not the whole PR. Just add whatever is coming in the commit and not the following commits.

nashif commented on subsys/debug/Kconfig:
```diff
@@ -43,6 +45,99 @@ config TRACING
 
+config TRACING_GENERIC
+	bool "Enabling Tracing generic infrastructure"
+	select TRACING
```
and enclose everything that depends on TRACING in an `if TRACING`

nashif commented:
> This is the first commit, but it is talking about the complete PR, please keep the commit messages to what the commit is doing and not the whole PR. Just add whatever is coming in the commit and not the following commits.

ignore.

wentongwu commented:
```
-:818: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
#818: FILE: subsys/debug/tracing/tracing_backend_posix.c:12:
+static FILE *out_stream;
             ^

- total: 1 errors, 1 warnings, 1619 lines checked
```
@nashif why `static FILE *out_stream;` is treated as error?

nashif commented on subsys/debug/tracing/tracing_backend_posix.c:
```diff
@@ -0,0 +1,38 @@
+static void tracing_backend_posix_init(void)
+{
+	out_stream =
+		(void *)fopen(CONFIG_TRACING_BACKEND_POSIX_FILE_NAME, "wb");
```
Why are you using a Kconfig here and not a command line option to the exe like it was done in subsys/debug/tracing/ctf/bottoms/posix/ctf_bottom.c?

wentongwu replied:
@nashif I think it's more easy to understand and save code size.

nashif replied:
I do not think it is easy to understand, it makes thing a bit more complex and incompatible with how it is done now, also, code size is not an issue with native_posix.

wentongwu replied:
@nashif you are right, we don’t need care about code size for native posix. And this   PR get file name at compile name, while current existing one gets posix file at run time, that’s the difference, I prefer the widely used Kconfig in Zephyr project.

wentongwu replied:
@nashif updated

dcpleung commented:
Currently, it looks like one can select multiple backends in kconfig, but the init function only uses one. It might be a good idea to wrap them under a menuconfig.

wentongwu commented:
> Currently, it looks like one can select multiple backends in kconfig, but the init function only uses one. It might be a good idea to wrap them under a menuconfig.

@dcpleung thanks, updated.

wentongwu committed:
tracing: add TRACING_ISR Kconfig

Add TRACING_ISR Kconfig to help high latency backend working well.

Currently the ISR tracing hook function is put at the begining and
ending of ISR wrapper, when there is ISR needed in the tracing path
(especially tracing backend), it will cause tracing buffer easily
be exhausted if async tracing method enabled. Also it will increase
system latency if all the ISRs are traced. So add TRACING_ISR to
enable/disable ISR tracing here. Later a filter out mechanism based
on irq number will be added.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
tracing: add TRACING_TEST Kconfig option

Add Kconfig option TRACING_TEST which can be used to
customize tracing packet format (currently supporting
string format and data format) for testing purpose.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
tracing: ctf: adapt ctf implementation to tracing infrastructure

Adapt ctf implementation to tracing infrastructure.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

nashif commented on tests/subsys/debug/tracing/CMakeLists.txt:
```diff
@@ -2,7 +2,17 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+if(BOARD MATCHES "mps2_.*")
```
why mps2? This should work with qemu_*. No need for a regex if we are going to target one platform...

wentongwu replied:
updated

nashif replied:
why not qemu_?

nashif commented:
in the sample, 
`west build -b qemu_x86 -t run  -- -DCONF_FILE="prj_uart.conf"` 

causes a crash.

nashif reviewed (changes requested):
> causes a crash.

@wentong this is a stack overlow, change add CONFIG_IDLE_STACK_SIZE=512 should fix this.
And change the "test" to run on all qemu please.

nslowell commented on subsys/debug/tracing/include/tracing_backend.h:
```diff
@@ -0,0 +1,108 @@
+	}
+
+	return NULL;
+}
```
It seems like there can only ever be one tracer? Are there future plans for multiple tracers at once?

If it's only one, can this function just blindly return backend?  If that's the case, is the name parameter still needed?

wentongwu replied:
I have no plan to support multiple tracers so far, because it would increase the system latency.

It chooses the backend based on the name

nslowell reviewed (commented):
Really like the added backends.  Have a few questions/comments.

wentongwu commented on subsys/debug/tracing/tracing_format_sync.c:
```diff
@@ -0,0 +1,75 @@
+		tracing_buffer_handle((u8_t *)data, length);
+		tracing_buffer_get_finish(length);
+	} else {
+		tracing_packet_drop_handle();
```
> Looks like 2 of the 3 are copying data into a buffer and immediately pulling it back out for handling simply because it needs to be formatted into "data" and "length" parameters.

yes, because tracing_buffer_handle need them.

 > Is there a different non-buffered method like tracing_format_raw_data()?

I just define 3 different APIs from which end users have to choose based on their usage cases. In this file they work in sync way.

wentongwu commented on subsys/debug/tracing/tracing_backend_uart.c:
```diff
@@ -0,0 +1,103 @@
+
+static struct device *dev;
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
```
```
config TRACING_HANDLE_HOST_CMD
        bool "Enable host cmd handle"
        select UART_INTERRUPT_DRIVEN if TRACING_BACKEND_UART
        help
          When enabled tracing will handle cmd from host to dynamically
          enable and disable tracing to have host capture tracing stream
          data conveniently.
```

nslowell replied:
It is possible for CONFIG_UART_INTERRUPT_DRIVEN to be enabled but CONFIG_TRACING_HANDLE_HOST_CMD to not be enabled.  If that's the case, this code isn't needed, so it should be based on CONFIG_TRACING_HANDLE_HOST_CMD.

wentongwu replied:
@nslowell ok, thanks. Updated.

wentongwu commented on subsys/debug/tracing/tracing_core.c:
```diff
@@ -0,0 +1,163 @@
+				tracing_backend_get(TRACING_BACKEND_POSIX_NAME);
+
+			tracing_backend_init(working_backend);
+		}
```
thanks

wentongwu commented on subsys/debug/tracing/tracing_core.c:
```diff
@@ -0,0 +1,163 @@
+
+#define TRACING_BACKEND_UART_NAME   "tracing_backend_uart"
+#define TRACING_BACKEND_USB_NAME    "tracing_backend_usb"
+#define TRACING_BACKEND_POSIX_NAME  "tracing_backend_posix"
```
thanks

wentongwu commented on subsys/debug/tracing/tracing_core.c:
```diff
@@ -0,0 +1,163 @@
+	if (before_put_is_empty) {
+		k_timer_start(&tracing_thread_timer,
+			      CONFIG_TRACING_THREAD_WAIT_THRESHOLD, K_NO_WAIT);
+	}
```
this is under ASYNC way, we need some time to do the buffer. If would like directly output tracing message, sync way is a good choice.

wentongwu commented on subsys/debug/tracing/tracing_backend_uart.c:
```diff
@@ -0,0 +1,103 @@
+{
+	for (u32_t i = 0; i < length; i++) {
+		uart_poll_out(dev, data[i]);
+	}
```
will consider this enhancement, thanks.

wentongwu commented on scripts/tracing/trace_capture_usb.py:
```diff
@@ -0,0 +1,103 @@
+    else:
+        product_id = int(args.product_id, 16)
+
+    output_file = args.output
```
will add default value, thanks

wentongwu commented on scripts/tracing/trace_capture_uart.py:
```diff
@@ -0,0 +1,59 @@
+    parse_args()
+    serial_port = args.serial_port
+    serial_baudrate = args.serial_baudrate
+    output_file = args.output
```
will add default value, thanks

wentongwu committed:
scripts: add script to capture tracing data with UART backend

Add script to capture tracing stream data with UART backend. This
script is developed based on pyserial, so install it correctly
before using the script.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
scripts: add script for usb backend to receive tracing data

Add script for usb backend to receive tracing stream data. This
script, trace_capture_usb.py, is based on pyusb, so install it
correctly before using the script.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu commented on subsys/debug/tracing/include/tracing_backend.h:
```diff
@@ -0,0 +1,108 @@
+		const struct tracing_backend *backend)
+{
+	if (backend) {
+		backend->api->init();
```
updated

wentongwu commented on subsys/debug/tracing/include/tracing_backend.h:
```diff
@@ -0,0 +1,108 @@
+		u8_t *data, u32_t length)
+{
+	if (backend) {
+		backend->api->output(backend, data, length);
```
@nslowell thanks, updated.

wentongwu commented:
> > causes a crash.
> 
> @wentong this is a stack overlow, change add CONFIG_IDLE_STACK_SIZE=512 should fix this.
> And change the "test" to run on all qemu please.

@nashif thanks.

wentongwu committed:
tests: tracing: add test cases for tracing infrastructure

Add test cases for currently supported backends (UART, USB
and POSIX) of tracing infrastructure.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

ceolin commented on subsys/debug/tracing/tracing_backend_posix.c:
```diff
@@ -0,0 +1,66 @@
+#include <tracing_backend.h>
+
+static void *out_stream;
+static const char *file_name;
```
How is supposed to set `file_name` ? The scope of this variable is this file and there is no API to change it.

wentongwu replied:
```
void tracing_backend_posix_option(void)
{
        static struct args_struct_t tracing_backend_option[] = {
                {
                        .manual = false,
                        .is_mandatory = false,
                        .is_switch = false,
                        .option = "trace-file",
                        .name = "file_name",
                        .type = 's',
                        .dest = (void *)&file_name,
                        .call_when_found = NULL,
                        .descript = "File name for tracing output.",
                },
                ARG_TABLE_ENDMARKER
        };

        native_add_command_line_opts(tracing_backend_option);
}

```
native_add_command_line_opts will parse tracing_backend_option which contains the address of file_name.

ceolin replied:
my point is, how the user change the value of `file_name` ?  The struct `tracing_backend_option` is statically defined with the address of `file_name` that is constant.

wentongwu replied:
@ceolin file_name will point to const string passed by user(when run ./zephyr.elf channel0_1), it can point another string(file name), but we don't have that case, all the tracing data will only be save one file.

ceolin replied:
got it !

ceolin commented on subsys/debug/tracing/tracing_backend_posix.c:
```diff
@@ -0,0 +1,66 @@
+#include <cmdline.h>
+#include <tracing_backend.h>
+
+static void *out_stream;
```
why not use `FILE *` and avoid casts ?

wentongwu replied:
```
-:818: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
#818: FILE: subsys/debug/tracing/tracing_backend_posix.c:12:
+static FILE *out_stream;
```
it seems checkpatch doesn't like this.

ceolin commented on subsys/debug/tracing/tracing_backend_uart.c:
```diff
@@ -0,0 +1,103 @@
+
+		if (!cmd) {
+			length = tracing_cmd_buffer_alloc(&cmd);
+			if (!cmd) {
```
you have checked `cmd` before. `tracing_cmd_buffer_alloc` never sets it to NULL, it always return the address of the first element of this static array that has at least one element.

wentongwu replied:
thanks, updated.

ceolin commented on subsys/debug/tracing/tracing_backend_uart.c:
```diff
@@ -0,0 +1,103 @@
+		}
+
+		if (!isprint(byte)) {
+			switch (byte) {
```
this switch could be an `if` since there is checking for only one possible case. (MISRA-C has rules related with it)

wentongwu replied:
thanks, updated

ceolin commented on subsys/debug/tracing/tracing_backend_uart.c:
```diff
@@ -0,0 +1,103 @@
+static void tracing_backend_uart_init(void)
+{
+	dev = device_get_binding(CONFIG_TRACING_BACKEND_UART_NAME);
+	assert(dev);
```
why don't you use `__ASSERT` that can be turned on/off, gives more information and can users can set a post-action function ?

wentongwu replied:
yes, should give more info if init failed. updated, thanks

ceolin commented on scripts/tracing/trace_capture_uart.py:
```diff
@@ -0,0 +1,59 @@
+    try:
+        main()
+    except KeyboardInterrupt:
+        print('Data capture interrupted, data saved into {}'.format(args.output))
```
shouldn't you  close the file descriptor or is it somehow automatically done ?

wentongwu replied:
https://stackoverflow.com/questions/17577137/do-files-get-closed-during-an-exception-exit

ceolin commented:
nice work !

ceolin commented on include/debug/tracing_format.h:
```diff
@@ -0,0 +1,79 @@
+ * @param data   Raw data to be traced.
+ * @param length Raw data length.
+ */
+void tracing_format_raw_data(const char *data, u32_t length);
```
it is really minor, but I would use `u8_t` instead of `char`. The C standard does not define if `char` is signed or unsigned, so for this case `u8_t` is just better and you already store it internally as `u8_t` anyway. For strings `char` is ok :)

wentongwu replied:
@ceolin thanks

wentongwu committed:
tracing: add infrastructure for collection of tracing data

First, this commit adds user interface in tracing_format.h which
can trace both string format and data format packet.

Second, it adds method both for asynchronous and synchronous way.
For asynchronous method, tracing packet will be buffered in tracing
buffer first, tracing thread will output the stream data with the
help of tracing backend when tracing thread get scheduled.

Third, it adds UART and USB tracing backend for asynchronous
tracing method, and adds POSIX tracing backend for synchronous
tracing way.

Also it can receive command from host to dynamically enable and
disable tracing to have host capture tracing data conveniently.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>
